File: Src/Runtime/FileSystem/FileName.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 6/15/26.
//

#include "FileName.h"

#include <filesystem>

#include "ADT/Exception/CouldNotCreateDirectory.h"
#include "Support/Runtime/AccessString.h"
#include "Support/Runtime/ConvertToString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::FileSystem;

ADT::Runtime::Object FileSystem::FileName(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &path_obj = args[0];
    auto path_ext = Support::Runtime::AccessString(path_obj);
    auto path_str = Support::Runtime::ConvertToString(path_ext);

    auto path = std::filesystem::path(path_str.CStr());
    auto res = path.filename();
    auto res_str = res.string();

    return {
        ADT::Runtime::ObjectType::String,
        Celery::Str::String(res_str.data(), res_str.size())
    };
}

File: src/runtime/file_system/file_name.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use std::path::Path;
use crate::adt::error::RuntimeError;
use crate::adt::lang::{ASTValue, RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::get_string;

pub fn file_name<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let path = args.get(0).unwrap();
    let path = get_string(path, trace)?;

    let res = Path::new(path);
    let stem = res.file_stem();
    if stem.is_none() {
        return Err(
            RuntimeError::could_not_read(trace)
        );
    }

    let stem = stem.unwrap();
    let stem = stem.to_str();

    if stem.is_none() {
        return Err(
            RuntimeError::could_not_read(trace)
        );
    }

    let stem = stem.unwrap();
    continue_execution(
        Object::Hashable(
            HashableObject::String(
                ASTValue::Owned(stem.to_string())
            )
        )
    )
}